### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by aspcompiler**
Initial commit of openpulse-python. It is based on [the previous PR to openqasm repo](https://github.com/openqasm/openqasm/pull/352). The main change is that we decoupled from OpenQASM parser and only parse `cal` and `defcal` body. Currently, we walk the AST from OpenQASM parser and parse the text body of `cal`/`defcal` with the OpenPulse parser. In the future, we will use [delegating parser](https://github.com/jakelishman/openqasm/tree/delegating-parser) when it is merged into the OpenQASM.

### CODE REVIEW

File: source/openpulse/openpulse/parser.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+OpenPulseNodeVisitor.visitResetStatement = QASMNodeVisitor.visitResetStatement
+OpenPulseNodeVisitor.visitScope = QASMNodeVisitor.visitScope
+OpenPulseNodeVisitor.visitSetExpression = QASMNodeVisitor.visitSetExpression
+OpenPulseNodeVisitor.visitStatement = QASMNodeVisitor.visitStatement
+OpenPulseNodeVisitor.visitStatementOrScope = QASMNodeVisitor.visitStatementOrScope
+OpenPulseNodeVisitor.visitUnaryExpression = QASMNodeVisitor.visitUnaryExpression
+OpenPulseNodeVisitor.visitWhileStatement = QASMNodeVisitor.visitWhileStatement
+
+
+class CalParser(QASMVisitor[None]):
```
Comment (aspcompiler) [inline source/openpulse/openpulse/parser.py:290]: Before [delegating-parser](https://github.com/jakelishman/openqasm/tree/delegating-parser) is merged into OpenQASM mainline, I used this Visitor to simulate the effect.

File: source/openpulse/openpulse/parser.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+    def visitRangeExpression(self, ctx: openpulseParser.RangeExpressionContext):
+        # start, end and step are all optional as in [:]
+        # It could be [start:end] or [start:step:end]
+        start = None
+        end = None
+        step = None
+        colons_seen = 0
+
+        for child in ctx.getChildren():
+            if isinstance(child, openpulseParser.ExpressionContext):
```
Comment (aspcompiler) [inline source/openpulse/openpulse/parser.py:174]: @jakelishman, I tried your [Meta hack](https://github.com/openqasm/openqasm/pull/352#discussion_r950325285) but it failed to pass the unit tests. Somehow the code generated by different versions of Antlr is significantly different; I am using version 4.9.2 which is the one used in OpenQASM git action. I tried different hacks but each of them only pass a subset of the unit tests.

**Comment by aspcompiler** [comment]
Somehow I was not able to request reviewers. Please take a look when you get an opportunity @jakelishman and @philreinhold. Thanks!

File: source/openpulse/openpulse/parser.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+    lexer = openpulseLexer(InputStream(input_))
+    stream = CommonTokenStream(lexer)
+    parser = openpulseParser(stream)
+    tree = parser.calibrationBlock()
+    result = (
+        OpenPulseNodeVisitor(in_defcal).visitCalibrationBlock(tree)
+        if tree.children
+        else openpulse_ast.CalibrationBlock(body=[])
+    )
+    return result
```
Comment (psivaraj) [inline source/openpulse/openpulse/parser.py:75]: **question**: I don't quite understand when we'd hit the `else` statement. Is this when there's an empty `cal` statement?

File: source/openpulse/openpulse/parser.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+# Reuse some QASMNodeVisitor methods in OpenPulseNodeVisitor
+# The following methods are overridden in OpenPulseNodeVisitor and thus not imported:
+"""
+    "visitChildren",
+    "visitErrorNode",
+    "visitTerminal",
+    "visitArrayLiteral",
+    "visitCalibrationGrammarStatement",
+    "visitCalStatement",
+    "visitDefcalStatement",
```
Comment (psivaraj) [inline source/openpulse/openpulse/parser.py:211]: **issue**: I don't see any of these overridden in `OpenPulseNodeVisitor`.

**Review by psivaraj — APPROVED** [review]

File: source/openpulse/openpulse/parser.py (hunk 3 above)
Comment (aspcompiler) [reply source/openpulse/openpulse/parser.py:75]: Exactly. Empty `cal` statement.

File: source/openpulse/openpulse/parser.py (hunk 4 above)
Comment (aspcompiler) [reply source/openpulse/openpulse/parser.py]: Indeed. Forgot to update the comments. Fixed.

**Pull request merged**